proposed/http-message.md

```diff
@@ -232,15 +236,184 @@ interface MessageInterface
+     * method names are case-sensitive and thus implementations SHOULD NOT
+     * modify the given string.
+     *
+     * @param string $method Case-insensitive method.
+     * @return void
+     * @throws \InvalidArgumentException for invalid HTTP methods.
+     */
+    public function setMethod($method);
+
+    /**
+     * Retrieves the absolute URI.
+     *
+     * An absolute URI consists of minimally scheme and host, but can also
+     * contain:
+     *
+     * - authentication (user/pass) if provided
+     * - port (if non-standard)
+     * - path (if any)
+     * - query string (if present)
+     * - fragment (if present)
+     *
+     * If either of the scheme or host are not present, this method MUST return
+     * null.
+     *
+     * @link http://tools.ietf.org/html/rfc3986#section-4.3
+     * @return string|null Returns the absolute URL as a string. The URL MUST
+     *     include the scheme and host; if the port is non-standard for the
+     *     scheme, the port MUST be included; authentication data MAY be
+     *     provided. If either host or scheme are missing, this method MUST
+     *     return null.
+     */
+    public function getAbsoluteUri();
+
+    /**
+     * Sets the absolute URI of the request.
+     *
+     * The absolute URI MUST be a string, and MUST include the scheme and host.
+     *
+     * If the port is non-standard for the scheme, the port MUST be provided.
+     *
+     * Authentication data MAY be provided.
+     *
+     * Path, query string, and fragment are optional.
+     *
+     * When setting the absolute URI, the url (see getUrl() and setUrl()) MUST
+     * be updated.
+     *
+     * @link http://tools.ietf.org/html/rfc3986#section-4.3
+     * @param string $uri Absolute request URI.
+     * @return void
+     * @throws \InvalidArgumentException If the URI is invalid.
+     */
+    public function setAbsoluteUri($uri);
+
+    /**
+     * Retrieves the request URL.
+     *
+     * The request URL is the path and query string ONLY.
+     *
+     * @link http://tools.ietf.org/html/rfc7230#section-5.3
+     * @return string Returns the URL as a string. The URL MUST be an
+     *     origin-form (path + query string), per RFC 7230 section 5.3
+     */
+    public function getUrl();
+
+    /**
+     * Sets the request URL.
+     *
+     * The URL MUST be a string. The URL SHOULD be an origin-form (path + query
+     * string) per RFC 7230 section 5.3; if other URL parts are present, the
+     * method MUST raise an exception OR remove those parts.
+     *
+     * When setting the URL, the absolute URI (see getAbsoluteUri() and
+     * setAbsoluteUri()) MUST be updated.
+     *
+     * @link http://tools.ietf.org/html/rfc7230#section-5.3
+     * @param string $url Request URL, with path and optionally query string.
+     * @return void
+     * @throws \InvalidArgumentException If the URL is invalid.
+     */
+    public function setUrl($url);
```

Comment:
The problem is that "request target" has 4 potential forms, ranging from least information (origin-form) to most (fully qualified URI that includes all of scheme, authentication, host, port, path, and query).

The "absolute URI" member essentially address the "most" target, while the "URL" member addresses the least. @evert has made the case that URL is a well-known subset of URI, and is used to specify origin-form -- but, as you note, it could easily be confused with URI, leading to odd error conditions or expectation problems.

`setUriOriginForm()` would be a head-scratcher for anybody who has not read the spec -- which would be a majority of users. As such, I think it's not a good candidate.

Perhaps we should do absolute URI and path only? My only issue with that is if the path changes, the query string likely need to change as well, and then we're back to needing getters/setters for every URI part, or at least a plethora of them.